Fix bugs that occur when two duplicate songs appear on feed from diff…

ChenJesse opened the pull request:
…erent users

#257

ChenJesse committed:
Fix bugs that occur when two duplicate songs appear on feed from different users

loganallen commented:
If it works, looks good to me

dennisfedorko commented on Tempo/Controllers/PlayerTableViewController.swift:
```diff
@@ -47,7 +47,10 @@ class PlayerTableViewController: UIViewController, UITableViewDelegate, UITableV
 				array = filteredPosts
 			}
-            if let row = currentlyPlayingIndexPath?.row, (playerNav.currentPost?.song.equals(other: array[row].song) ?? false && self == (playerNav.playerCell.delegate as! PlayerTableViewController)) {
+            if let row = currentlyPlayingIndexPath?.row,
```
you should unwrap currentPost so you don't have to do this ?? nonsense

dennisfedorko commented on Tempo/Controllers/PlayerTableViewController.swift:
```diff
@@ -47,7 +47,10 @@ class PlayerTableViewController: UIViewController, UITableViewDelegate, UITableV
+            if let row = currentlyPlayingIndexPath?.row,
+				(playerNav.currentPost?.song.equals(other: array[row].song) ?? false
+				&& (playerNav.currentPost?.user.equals(other: array[row].user)) ?? false
+				&& self == (playerNav.playerCell.delegate as! PlayerTableViewController)) {
```
dont force cast here, you should optionally cast and unwrap using if let

dennisfedorko commented on Tempo/Models/User.swift:
```diff
@@ -112,6 +112,10 @@ class User: NSObject, NSCoding {
 		aCoder.encode(username, forKey: "username")
 	}
+	
+	func equals(other: User) -> Bool {
```
doesn't each user have a unique identifier, isn't that a better than using username?

ChenJesse replied:
I don't think so

dennisfedorko replied:
We actually have 3 unique identifiers on a user object: id, fbid, and username

ChenJesse replied:
Lol my b, idk how I missed that...

ChenJesse commented:
Ok I thought of a slightly better way to organize the checks, I'll get it done by tonight

say25 commented:
SONG ID?

ChenJesse commented:
What do you mean?

say25 commented:
Songs have a unique Spotify given ID on them. If a user posts truly the same song it by definition has the same song id. that is probably the best way to determine duplicates

ChenJesse commented:
Oh, the problem isn't prevent different users from posting the same song, it's dealing with how the logic determines which song is playing when two of them are identical. I do what you're talking about when I do song1.equals(other: song2), which compares IDs. But that alone is not sufficient for when two people you follow post the same song (with same songIDs, since the one further up in the tableview will always get played first)

dennisfedorko commented:
I guess using username is about the same as id, so works for me

ChenJesse committed:
Fix more feed bugs and fix up style

ChenJesse commented:
Ok, fixed up the style, using id for comparing Users, and now defines Post equality as having the same equal users and songs.

Also I found a few other bugs and fixed them:
- Scrolling away from a spinning feedCell and coming back would sometimes stop animation
- Pull to refresh animation continuity of playing song was inconsistent, moved the function call to a better place

dennisfedorko commented:
Merge it
